core/src/main/java/org/apache/iceberg/encryption/Ciphers.java

```diff
@@ -101,37 +117,52 @@ public AesGcmDecryptor(byte[] keyBytes) {
     }
 
     public byte[] decrypt(byte[] ciphertext, byte[] aad) {
-      int plainTextLength = ciphertext.length - GCM_TAG_LENGTH - NONCE_LENGTH;
+      return decrypt(ciphertext, 0, ciphertext.length, aad);
+    }
+
+    public byte[] decrypt(
+        byte[] ciphertext, int ciphertextOffset, int ciphertextLength, byte[] aad) {
       Preconditions.checkState(
-          plainTextLength >= 1,
+          ciphertextLength - GCM_TAG_LENGTH - NONCE_LENGTH >= 1,
           "Cannot decrypt cipher text of length "
               + ciphertext.length
               + " because text must longer than GCM_TAG_LENGTH + NONCE_LENGTH bytes. Text may not be encrypted"
               + " with AES GCM cipher");
 
       // Get the nonce from ciphertext
       byte[] nonce = new byte[NONCE_LENGTH];
-      System.arraycopy(ciphertext, 0, nonce, 0, NONCE_LENGTH);
+      System.arraycopy(ciphertext, ciphertextOffset, nonce, 0, NONCE_LENGTH);
```

Comment:
The allocation here is unnecessary. `GCMParameterSpec` accepts an offset and length for the nonce array.

Reply:
Cool!